docs/docs/developers/docs/aztec-nr/framework-description/state-variables.md

```diff
@@ -0,0 +1,558 @@
+#### `initialize`
+
+When this function is invoked, it creates a nullifier for the storage slot, ensuring that the `PrivateImmutable` cannot be initialized again:
+
+```rust
+#[external("private")]
+fn initialize_key(key_value: Field) {
+    let owner = self.msg_sender();
+    let note = KeyNote::new(key_value, owner);
+    self.storage.signing_key.at(owner).initialize(note).emit(owner, MessageDelivery.CONSTRAINED_ONCHAIN);
+}
+```
+
+#### `get_note`
+
+Similar to the `PrivateMutable`, we can use the `get_note` method to read the value:
+
+```rust
+#[external("private")]
+fn get_key() -> KeyNote {
+    let owner = self.msg_sender();
+    self.storage.signing_key.at(owner).get_note()
+}
+```
+
+Unlike a `PrivateMutable`, the `get_note` function for a `PrivateImmutable` doesn't nullify the current note. This means that multiple accounts can concurrently call this function to read the value.
+
+### PrivateSet
+
+`PrivateSet` is used for managing a collection of notes. Like `PrivateMutable`, this is a private state variable that can be modified. There are two key differences:
+
+- A `PrivateSet` is not a single value but a _set_ (a collection) of values (represented by notes)
+- Any account can insert values into someone else's set.
+
+The set's current value is the collection of notes in the set that have not yet been nullified. These notes can have any type: they could be nft IDs, representing a user's nft collection, or they might be token amounts, in which case _the sum_ of all values in the set would be the user's current balance.
+
+#### Declaration
+
+For example, to add a mapping of private token balances to storage:
+
+```rust
+#[storage]
+struct Storage<Context> {
+    balances: Map<AztecAddress, PrivateSet<UintNote, Context>, Context>,
+}
+```
+
+#### `insert`
+
+Allows us to modify the storage by inserting a note into the `PrivateSet`:
+
+```rust
+#[external("private")]
+fn mint_tokens(to: AztecAddress, amount: u128) {
+    let note = UintNote::new(amount, to);
+    self.storage.balances.at(to).insert(note).emit(to, MessageDelivery.UNCONSTRAINED_ONCHAIN);
+}
+```
+
+#### `get_notes`
+
+Retrieves notes the account has access to. You can optionally provide filtering options:
+
+```rust
+// Get all notes (with default options)
+let options = NoteGetterOptions::new();
+let notes = self.storage.balances.at(owner).get_notes(options);
+
+// Or with custom options (e.g., limit the number of notes)
+let options = NoteGetterOptions::new().set_limit(5);
+let notes = self.storage.balances.at(owner).get_notes(options);
+```
+
+#### `pop_notes`
+
+This function pops (gets, removes and returns) the notes the account has access to:
+
+```rust
+// Pop notes with a limit
+let options = NoteGetterOptions::new().set_limit(10);
+let notes = self.storage.balances.at(owner).pop_notes(options);
+```
+
+#### `remove`
+
+Will remove a note from the `PrivateSet` if it previously has been read from storage:
+
+```rust
+self.storage.balances.at(owner).remove(note);
```

Comment:
`.remove` takes `RetrievedNote<Note>`. whats the way to get `RetrievedNote<Note>`?